internal/graph/check.go

```diff
@@ -602,7 +586,22 @@ func (c *LocalChecker) checkDirect(parentctx context.Context, req *ResolveCheckR
 			return union(ctx, c.concurrencyLimit, handlers...)
 		}
 
-		checkFuncs = append(checkFuncs, fn2)
+		var checkFuncs []CheckHandlerFunc
+
+		shouldCheckDirectTuple, _ := typesys.IsDirectlyRelated(
+			typesystem.DirectRelationReference(objectType, relation),                                         //target
+			typesystem.DirectRelationReference(tuple.GetType(tk.GetUser()), tuple.GetRelation(tk.GetUser())), //source
+		)
+
+		if shouldCheckDirectTuple {
+			checkFuncs = []CheckHandlerFunc{fn1}
+		}
+
+		directlyRelatedUserTypesHasUsersetTuples, _ := typesys.DirectlyRelatedUserTypesHasUsersetTuples(objectType, relation)
+		if directlyRelatedUserTypesHasUsersetTuples {
+			checkFuncs = append(checkFuncs, fn2)
```

Comment:
This change is good because if the Check is done over, for example, `define viewer: [user, employee]`, we will not call fn2.

However, it's not enough... if you have a mixture of users and usersets, e.g. `define viewer: [user, group#member]`, then inside code `fn2` we are still sending `user` in the call to `ReadUsersetTuples` which is inefficient because `user` is not a userset.

I would change this code to

```go
listOfUsersets := typesys.DirectlyRelatedUserTypesThatAreUsersets(objectType, relation)
if len(listOfUsersets) > 0 {
	checkFuncs = append(checkFuncs, fn2)
}
```

And then inside `fn2` change 505 to be

```go
allowedUserTypeRestrictions := typesys.DirectlyRelatedUserTypesThatAreUsersets(objectType, relation)
```

Reply:
Also, i'm not a fan of this code being outside `fn1` and `fn2` when really, they belong there.

How about we have

```go
var checkFuncs = []CheckHandlerFunc{fn1, fn2}
return union(ctx, c.concurrencyLimit, checkFuncs...)
```

and we move these optimization checks within each fn?

```go
fn1 := func(ctx context.Context) (*ResolveCheckResponse, error) {
			ctx, span := tracer.Start(ctx, "checkDirectUserTuple", trace.WithAttributes(attribute.String("tuple_key", tk.String())))
			defer span.End()

			response := &ResolveCheckResponse{
				Allowed: false,
				ResolutionMetadata: &ResolutionMetadata{
					DatastoreQueryCount: req.GetResolutionMetadata().DatastoreQueryCount + 1,
				},
			}

			shouldCheckDirectTuple, _ := typesys.IsDirectlyRelated(
				typesystem.DirectRelationReference(objectType, relation),                                         //target
				typesystem.DirectRelationReference(tuple.GetType(tk.GetUser()), tuple.GetRelation(tk.GetUser())), //source
			)

			if !shouldCheckDirectTuple {
				return response, nil
			}
```

and

```go
fn2 := func(ctx context.Context) (*ResolveCheckResponse, error) {
			ctx, span := tracer.Start(ctx, "checkDirectUsersetTuples", trace.WithAttributes(attribute.String("userset", tuple.ToObjectRelationString(tk.Object, tk.Relation))))
			defer span.End()

			response := &ResolveCheckResponse{
				Allowed: false,
				ResolutionMetadata: &ResolutionMetadata{
					DatastoreQueryCount: req.GetResolutionMetadata().DatastoreQueryCount + 1,
				},
			}

			allowedUserTypeRestrictions := typesys.DirectlyRelatedUserTypesThatAreUsersets(objectType, relation)
			if len(allowedUserTypeRestrictions) == 0 {
				return response, nil
			}
```

Reply:
You needlessly launch goroutines, and memory allocations, empty spans instead of short circuiting early.

Reply:
> Also, i'm not a fan of this code being outside fn1 and fn2 when really, they belong there.

As noted by @elbuo8 , we do not want to launch the GoRoutine itself as it responds more slowly + require go routines allocation.

Reply:
> That ends up launching goroutines, memory allocations

This is true, but we gain in readability because these checks happen where they are needed

> empty spans

How so? The code would be included in span `checkDirectUsersetTuples` which should be short-lived in these cases

Reply:
> This is true, but so far the bottlenecks we've seen are only about DB connections, no?

We had seen evidence in the past where some "large/complex" check requests causes GoRoutine / Garbage collection spikes because of the huge number of concurrent DB requests.  If there is any way to reduce load on the system, why do we want to do it otherwise?

Reply:
> Also, i'm not a fan of this code being outside fn1 and fn2 when really, they belong there.

We have no need to resolve fn1 and/or fn2 if there is not a type restriction that would lead to a resolution path. It's wasted cycles and GC and memory etc..

My vote here is to move this code outside and queue up/append the function handlers IFF they are needed based on the type restrictions.

```go
fn1 := func(...)
fn2 := func(...)

var checkFuncs = []CheckHandlerFunc{fn1, fn2}
if (fn1Condition) {
  checkFuncs = append(checkFuncs, fn1)
}

if (fn2Condition) {
  checkFuncs = append(checkFuncs, fn2)
}

return union(..., checkFuncs...)
```

Reply:
@adriantam sorry, I was reading the diff wrong. That's what we're doing below `fn2`. 👍